server/src/main/java/io/druid/curator/discovery/CuratorDruidLeaderSelector.java

```diff
@@ -0,0 +1,201 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.curator.discovery;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import com.metamx.emitter.EmittingLogger;
+import io.druid.concurrent.Execs;
+import io.druid.concurrent.LifecycleLock;
+import io.druid.discovery.DruidLeaderSelector;
+import io.druid.guice.annotations.Self;
+import io.druid.java.util.common.ISE;
+import io.druid.java.util.common.StringUtils;
+import io.druid.java.util.common.guava.CloseQuietly;
+import io.druid.server.DruidNode;
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.recipes.leader.LeaderLatch;
+import org.apache.curator.framework.recipes.leader.LeaderLatchListener;
+import org.apache.curator.framework.recipes.leader.Participant;
+
+import javax.annotation.Nullable;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ */
+public class CuratorDruidLeaderSelector implements DruidLeaderSelector
+{
+  private static final EmittingLogger log = new EmittingLogger(CuratorDruidLeaderSelector.class);
+
+  private final LifecycleLock lifecycleLock = new LifecycleLock();
+
+  private final DruidNode self;
+  private final CuratorFramework curator;
+  private final String latchPath;
+
+  private DruidLeaderSelector.Listener listener = null;
+  private final AtomicReference<LeaderLatch> leaderLatch = new AtomicReference<>();
+
+  private volatile boolean leader = false;
+  private volatile int term = 0;
+
+  public CuratorDruidLeaderSelector(CuratorFramework curator, @Self DruidNode self, String latchPath)
+  {
+    this.curator = curator;
+    this.self = self;
+    this.latchPath = latchPath;
+  }
+
+  private LeaderLatch createNewLeaderLatch()
+  {
+    final LeaderLatch newLeaderLatch = new LeaderLatch(
+        curator, latchPath, self.getHostAndPortToUse()
+    );
+
+    newLeaderLatch.addListener(
+        new LeaderLatchListener()
+        {
+          @Override
+          public void isLeader()
+          {
+            try {
+              if (leader) {
+                log.warn("I'm being asked to become leader. But I am already the leader. Ignored event.");
+                return;
+              }
+
+              leader = true;
+              term++;
+              listener.becomeLeader();
+            }
+            catch (Exception ex) {
+              log.makeAlert(ex, "listener becomeLeader() failed. Unable to become leader").emit();
+
+              // give others a chance to become leader.
+              final LeaderLatch oldLatch = createNewLeaderLatch();
```

Comment:
This is a really interesting way of doing recursion, but seems to match what the code was doing previously.

Reply:
yeah, i pretty much took the same code